flips/20211129-cadence-mutability-restrictions.md

```diff
@@ -0,0 +1,417 @@
+Of note is that the `let` kind does not allow fields to be written to in any scope, whereas `var` allows them
+to be written in the "Current and Inner" scopes; that is, the scope in which the field was declared, and any scopes
+contained within that scope. 
+
+However, simply writing to a field directly is not the only way in which one can modify a value. Consider the following example:
+
+```
+pub struct Foo {
+    pub let x : [Int]
+
+    init() {
+        self.x = [3];
+    }
+}
+
+pub fun bar() {
+    let foo = Foo()
+    foo.x = [0] // writes to x, not allowed
+    foo.x[0] = 0; // does not write to x, also not allowed
+}
+```
+
+Cadence also restricts the scopes in which an array or dictionary field can be modified (or "mutated"). Examples of 
+mutating operations include an indexed assignment, as in the above example, as well as calls to the `append` or `remove`,
+methods of arrays, or the `insert` or `remove` methods of dictionaries. These operations can only be performed on a field
+in the current and inner scopes, the same contexts in which the field could be written to if it were a `var`. So the following 
+would typecheck:
+
+```
+pub struct Foo {
+    pub let x : [Int]
+
+    init() {
+        self.x = [3];
+    }
+
+    pub fun addToX(i: Int) {
+        self.x.append(i)
+    }
+}
+```
+
+while the following would not:
+
+```
+pub struct Foo {
+    pub let y : [Int]
+
+    init() {
+        self.y = [3];
+    }
+}
+
+pub fun addToY(foo: Foo, i: Int) {
+    foo.y.append(i)
+}
+```
+
+This prevents external code from mutating the values of fields it can read from your contract. Consumers
+of your contract may read the values in a `pub let` or `pub var` field, but cannot change them in any way. 
+
+If you wish to allow other code to update or modify a field in your contract, you may expose a method 
+to do so, like in the example above with `addToX`, or you may use the `pub(set)` access mode, which 
+allows any code to mutate or write to the field it applies to. 
+
+### Implementation Details 
+
+This change adds a new error, the `ExternalMutationError`, which is raised when a field 
+is mutated outside of the context in which it was defined. The error message will also
+suggest that the user instead use a setter or modifier method for that field.
+
+Specifically, the error is emitted whenever a user attempts to perform an 
+index assignment on a member that is not either declared with the `pub(set)` 
+access mode, or is defined in the current enclosing scope. This check is the
+same one performed for writing to fields, with the difference that mutation 
+is allowed on both `let` and `var` fields, while only the latter can be written to.
+
+Additionally, array and dictionary methods now track an additional bit of information
+indicating whether they mutate their receiver. Mutating methods may not be called on 
+members that are not declared with the `pub(set)` access mode, or defined in the current
+enclosing scope. 
```

Comment:
Would we want to expose this to Cadence programs? Swift has the `mutating` function modifier for this purpose.

Just curious, I don't think that is necessary for this proposal, it could be added through a future proposal.

Reply:
We could; unsure what the use case is for this, but it's possible.